cpp/include/cuvs/neighbors/ball_cover.hpp

```diff
@@ -0,0 +1,352 @@
+  {
+    return R.view();
+  }
+  auto get_R_closest_landmark_dists() const -> raft::device_vector_view<const float, int64_t>
+  {
+    return R_closest_landmark_dists.view();
+  }
+  auto get_X_reordered() const -> raft::device_matrix_view<const float, int64_t, raft::row_major>
+  {
+    return X_reordered.view();
+  }
+
+  raft::device_vector_view<idx_t, int64_t> get_R_indptr() { return R_indptr.view(); }
+  raft::device_vector_view<idx_t, int64_t> get_R_1nn_cols() { return R_1nn_cols.view(); }
+  raft::device_vector_view<float, int64_t> get_R_1nn_dists() { return R_1nn_dists.view(); }
+  raft::device_vector_view<float, int64_t> get_R_radius() { return R_radius.view(); }
+  raft::device_matrix_view<float, int64_t, raft::row_major> get_R() { return R.view(); }
+  raft::device_vector_view<float, int64_t> get_R_closest_landmark_dists()
+  {
+    return R_closest_landmark_dists.view();
+  }
+  raft::device_matrix_view<float, int64_t, raft::row_major> get_X_reordered()
+  {
+    return X_reordered.view();
+  }
+  raft::device_matrix_view<const float, int64_t, raft::row_major> get_X() const { return X; }
+
+  cuvs::distance::DistanceType get_metric() const { return metric; }
+
+  int get_n_landmarks() const { return n_landmarks; }
+  bool is_index_trained() const { return index_trained; };
+
+  // This should only be set by internal functions
+  void set_index_trained() { index_trained = true; }
+
+  raft::resources const& handle;
+
+  int64_t m;
+  int64_t n;
+  int64_t n_landmarks;
+
+  raft::device_matrix_view<const float, idx_t, raft::row_major> X;
+
+  cuvs::distance::DistanceType metric;
+
+ private:
+  // CSR storing the neighborhoods for each data point
+  raft::device_vector<idx_t, int64_t> R_indptr;
+  raft::device_vector<idx_t, int64_t> R_1nn_cols;
+  raft::device_vector<float, int64_t> R_1nn_dists;
+  raft::device_vector<float, int64_t> R_closest_landmark_dists;
+
+  raft::device_vector<float, int64_t> R_radius;
+
+  raft::device_matrix<float, int64_t, raft::row_major> R;
+  raft::device_matrix<float, int64_t, raft::row_major> X_reordered;
+
+ protected:
+  bool index_trained;
+};
+
+/** @} */
+
+/**
+ * @defgroup random_ball_cover Random Ball Cover algorithm
+ * @{
+ */
+
+/**
+ * Builds and populates a previously unbuilt cuvs::neighbors::ball_cover::index
+ *
+ * Usage example:
+ * @code{.cpp}
+ *
+ *  #include <raft/core/resources.hpp>
+ *  #include <cuvs/neighbors/ball_cover.hpp>
+ *  #include <cuvs/distance/distance.hpp>
+ *  using namespace cuvs::neighbors;
+ *
+ *  raft::resources handle;
+ *  ...
+ *  auto metric = cuvs::distance::DistanceType::L2Expanded;
+ *  ball_cover::index index(handle, X, metric);
+ *  ball_cover::build_index(handle, index);
+ * @endcode
+ *
+ * @param[in] handle library resource management handle
+ * @param[inout] index an empty (and not previous built) instance of
+ * cuvs::neighbors::ball_cover::index
+ */
+void build(raft::resources const& handle, index<int64_t, float>& index);
```

Comment:
Just a note- we should have the factory function here too. We can do as a follow-on at some point. RBC isn't a heavily used API today.